### DISCUSSION TITLE
How to declare and use dependencies beyond de used framework (in the domain layer)

### DISCUSSION THREAD

**Discussion by javierseixas** [hynek/svcs#86] [Q&A]
Hello there!
We are building our application over Fast-api, for the external communication. However, we're applying Hexagonal Architecture, and we are developing some classes in our domain layer. This domain should be agnostic from the used framework, and we have domain classes that depend on other domain classes.
Here is where I become confused about how to use svcs, since all the samples relies on the framework. I come from other languages and frameworks, such as Java and Spring, and here Spring manages all the dependency tree from the app booting.
Do svcs allow this kind of services/dependencies management? How? Could I have any example?
Thanks for you help!

**Comment by hynek** [comment]
_svcs_ can do that for you too. If you want to use those dependencies outside the web framework in a web framework-agnostic way, simply a svcs container into your biz logic. Your biz logic would therefore do classic service location.

I personally am not a fan of that: I tend to acquire the services I need in the web layer and pass the instances into my biz logic (which makes it _dependency injection_) to decouple my business code from dependency management. But it’s up to you, what way you prefer.

I’m currently traveling and on a phone, so I can’t type you up an example, but I talk about these things quite a bit in my first two YouTube videos: 

- svcs: Robust & TESTABLE Python Web Applications with ANY Framework!
https://youtu.be/d1elMD9WgpA

- Loose Coupling & Dependency Injection the EASY Way!
https://youtu.be/uWTvMCra-_Y

**Reply by javierseixas** [reply]
Hi @hynek , thanks for your answer. My question could be a bit confusing, since I was (and still I'm) a bit lost. Let's rely the dependency management in the framework in the web layer, as you mention. With that in hand, I'm still a bit lost, in part because all the examples I've seen so far are creating a DB Connection. It's not my case yet. Let me explain: I want to apply composition into my code. Something like a decorator pattern. I have 2 classes and one interface:

```python
from abc import ABC, abstractmethod


class Dependency(ABC):

    @abstractmethod
    def speak(self):
        pass
```

And its implementations:

```python
class Decorator(Dependency):

    dependency: Decorated

    def __init__(self, dependency: Decorated):
        self.dependency = dependency

    def speak(self):
        print("Hello from dependency one")
        self.dependency.speak()

        return None
```
```python
class Decorated(Dependency):
    def speak(self):
        print("Hello from Decorated")
```

Having this, I've tried to register both implementation in svcs, but I couldn't since I got an error missing the factory method.
For solving that, I created a `create()` method in each implementation, and now FastAPI gets built. My services now look like:

```python
class Decorator(Dependency):

    dependency: Decorated

    @classmethod
    def create(cls, dependency: Decorated):
        return cls(dependency)

    def __init__(self, dependency: Decorated):
        self.dependency = dependency

    def speak(self):
        print("Hello from dependency one")
        self.dependency.speak()

        return None
```
```python
class Decorated(Dependency):
    @classmethod
    def create(cls, **kwargs):
        return cls(**kwargs)

    def speak(self):
        print("Hello from Decorated")
```

By the way, my setup for FastAPI looks like this:

```python
@svcs.fastapi.lifespan
async def lifespan(app: FastAPI, registry: svcs.Registry):
    registry.register_factory(Decorator, Decorator.create)
    registry.register_factory(Decorated, Decorated.create)

app = FastAPI(title=settings.PROJECT_NAME, lifespan=lifespan)
```

Now when I run the app, I get the following error: `TypeError("Decorator.create() missing 1 required positional argument: 'dependency'")`. The behaviour I expect is that svcs will be able to resolve that Decorator needs a Decorated instance for working, and pass it through the factory method.

Am I missing something?

Thanks for your help! :raised_hands: 

By the way: Your project looks very promising. I wish it gets a place in the python ecosystem. I hope to contribute to it by asking questions that may improve its acquisition by other engineers.

**Reply by hynek** [reply]
Hi Javier, sorry for the delay, I’ve been traveling and am only slowly catching up.

Your mental mismatch is that *svcs* is emphatically non-magical. It will **not** check the argument types of your factories and start building graphs from it or something.

That doesn’t mean that your use-case is impossible though, it just means, that you have to be explicit about it. As mentioned [here](https://svcs.hynek.me/en/stable/core-concepts.html#containers):

> If a factory takes a first argument called svcs_container or the first argument (of any name) is annotated as being [svcs.Container](https://svcs.hynek.me/en/stable/core-concepts.html#svcs.Container), the current container instance is passed into the factory as the first positional argument allowing for recursive service acquisition

In other words, you can do something like this (and I do it all the time):

```python

registry.register_factory(Decorated, Decorated) # doesn’t seem to take any arguments — `__init__` is perfectly adequate as a factory
registry.register_factory(Decorator, lambda svcs_container: Decorator(svcs_container.get(Decorated)))
```

Hope this makes sense and clears things up! Of course, you can use a classmethod factory on Decorator instead of a lambda, but I prefer to not couple my dependencies with *svcs*.

**edit**: also see #94
